src/mame/sinclair/pentevo.cpp

```diff
@@ -0,0 +1,693 @@
+		, m_glukrs(*this, "glukrs")
+		, m_sdcard(*this, "sdcard")
+		, m_keyboard(*this, "pc_keyboard")
+		, m_io_mouse(*this, "mouse_input%u", 1U)
+	{ }
+
+	void pentevo(machine_config &config);
+
+protected:
+	void machine_start() override;
+	void machine_reset() override;
+	void video_start() override;
+
+private:
+	void init_mem_write();
+	void pentevo_io(address_map &map);
+
+	u8 nemo_ata_r(u8 cmd);
+	void nemo_ata_w(u8 cmd, u8 data);
+
+	void atm_port_ffff_w(offs_t offset, u8 data) override;
+	void pentevo_port_f7f7_w(offs_t offset, u8 data);
+	void pentevo_port_fbf7_w(offs_t offset, u8 data);
+	void pentevo_port_eff7_w(offs_t offset, u8 data);
+	u8 atm_port_bf_r(offs_t offset);
+	void atm_port_bf_w(offs_t offset, u8 data);
+	u8 atm_port_0fbd_r(offs_t offset);
+	void atm_port_0fbd_w(offs_t offset, u8 data);
+	u8 atm_port_1fbd_r(offs_t offset);
+	void atm_port_1fbd_w(offs_t offset, u8 data);
+	u8 spi_port_77_r(offs_t offset);
+	void spi_port_77_w(offs_t offset, u8 data);
+	u8 spi_port_57_r(offs_t offset);
+	void spi_port_57_w(offs_t offset, u8 data);
+	void spi_miso_w(u8 data);
+
+	u8 gluk_data_r(offs_t offset);
+	void gluk_data_w(offs_t offset, u8 data);
+
+	TIMER_DEVICE_CALLBACK_MEMBER(nmi_check_callback);
+	void nmi_on();
+
+	bool is_shadow_active() override { return BIT(m_port_bf_data, 0) || is_dos_active(); }
+	bool is_gluk_active() {return is_shadow_active() || BIT(m_port_eff7_data, 7); }
+	bool is_port_7ffd_locked() override { return !is_pent1024() && BIT(m_port_7ffd_data, 5); }
+	bool is_pent1024() { return !BIT(m_port_eff7_data, 2); }
+
+	void atm_update_cpu() override;
+	u8 merge_ram_with_7ffd(u8 ram_page) override;
+
+	void spectrum_update_screen(screen_device &screen, bitmap_ind16 &bitmap, const rectangle &cliprect) override;
+	void pentevo_update_screen_zxhw(screen_device &screen, bitmap_ind16 &bitmap, const rectangle &cliprect);
+	void pentevo_update_screen_zx16(screen_device &screen, bitmap_ind16 &bitmap, const rectangle &cliprect);
+	void pentevo_update_screen_tx(screen_device &screen, bitmap_ind16 &bitmap, const rectangle &cliprect);
+
+	required_device<gfxdecode_device> m_gfxdecode;
+	required_device<ram_device> m_char_ram;
+
+	required_device<glukrs_device> m_glukrs;
+	required_device<spi_sdcard_sdhc_device> m_sdcard;
+	required_device<at_keyboard_device> m_keyboard;
+	required_ioport_array<3> m_io_mouse;
+
+	u8 m_port_bf_data;
+	u8 m_port_eff7_data;
+
+	u8 m_gluk_ext;
+	bool m_ata_data_hi_ready;
+	bool m_nmi_exit;
+
+	u8 m_zctl_di = 0;
+	u8 m_zctl_cs = 0;
+};
+
+
+/******************************************************************************
+ * ZX Evolution: BASECONF
+ * ***************************************************************************/
+
+void pentevo_state::atm_update_cpu()
+{
+	u8 multiplier = BIT(m_port_ff77_data, 3) ? 4 : (2 - BIT(m_port_eff7_data, 4));
+	m_maincpu->set_clock(X1_128_SINCLAIR / 10 * multiplier);
+}
+
+void pentevo_state::atm_port_ffff_w(offs_t offset, u8 data)
+{
+	if(!is_shadow_active())
```

Comment:
Please be consistent with spacing you’re formatted your other control statements like `if (foo)` (space before the opening parenthesis).